packages/firestore/test/integration/api/transactions.test.ts

```diff
@@ -23,7 +23,184 @@ import * as integrationHelpers from '../util/helpers';
 
 const apiDescribe = integrationHelpers.apiDescribe;
 
-apiDescribe('Database transactions', (persistence: boolean) => {
+apiDescribe.only('Database transactions', (persistence: boolean) => {
+  type TransactionStep = (
+    transaction: firestore.Transaction,
+    docRef: firestore.DocumentReference
+  ) => void;
+
+  /**
+   * Used for testing that all possible combinations of executing transactions
+   * result in the desired document value or error.
+   *
+   * The transaction steps are postfixed by numbers to indicate the calling
+   * order. For example, calling `set1()` followed by `set2()` should result in
+   * the document being set to the value specified by `set2()`.
+   */
+  class TransactionTester {
+    constructor(readonly db: firestore.FirebaseFirestore) {}
+
+    async delete(
+      transaction: firestore.Transaction,
+      docRef: firestore.DocumentReference
+    ): Promise<void> {
+      transaction.delete(docRef);
+    }
+
+    async update1(
+      transaction: firestore.Transaction,
+      docRef: firestore.DocumentReference
+    ): Promise<void> {
+      transaction.update(docRef, { foo: 'bar1' });
+    }
+
+    async update2(
+      transaction: firestore.Transaction,
+      docRef: firestore.DocumentReference
+    ): Promise<void> {
+      transaction.update(docRef, { foo: 'bar2' });
+    }
+
+    async set1(
+      transaction: firestore.Transaction,
+      docRef: firestore.DocumentReference
+    ): Promise<void> {
+      transaction.set(docRef, { foo: 'bar1' });
+    }
+
+    async set2(
+      transaction: firestore.Transaction,
+      docRef: firestore.DocumentReference
+    ): Promise<void> {
+      transaction.set(docRef, { foo: 'bar2' });
+    }
+
+    async getExisting(
+      transaction: firestore.Transaction,
+      docRef: firestore.DocumentReference
+    ): Promise<void> {
+      return transaction.get(docRef).then(docSnap => {
+        expect(docSnap.exists).to.equal(true);
+      });
+    }
+
+    async getMissing(
+      transaction: firestore.Transaction,
+      docRef: firestore.DocumentReference
+    ): Promise<void> {
+      return transaction.get(docRef).then(docSnap => {
+        expect(docSnap.exists).to.equal(false);
+      });
+    }
+
+    /**
+     * Validates that executing the provided steps in a transaction results in
+     * the doc value equal to the expected value.
+     *
+     * @param expected The expected value of the doc.
+     * @param docExists Whether the document the transaction is being run on
+     * should exist beforehand.
+     * @param stages The series of transaction steps to execute inside the
+     * transaction.
+     */
+    async validateSequenceSuccess(
+      expected: object | undefined,
```

Comment:
The ordering of these arguments ends up reading weirdly.

First is the expected outcome, then the precondition for the test, then the actual steps. You can't reorder the varargs, but you could make the varargs `TransactionStep | object | undefined` and then pop the last one off as the actual expected outcome.

Reply:
I originally implemented it with popping off the last arg, but I realized that if this case where you get an existing document, the document has to be set outside of the `runTransaction` block, which means you can't run it as a `TransactionStage`. I think popping off the last value and the 2nd to last value to do different things is worse for readability compared to having a different order that have comments in the input. Thoughts?

Reply:
We have to balance readability in the caller vs readability in the implementation. Commenting without reordering is a sort of band-aid. It's fine, I guess, but I think we can do better.

This would read better in the caller of we could make it read like: establish precondition, run stages, verify result.

Some alternatives:

  * Take an explicit array of steps as a middle argument rather than varargs
  * Define overloads that enumerate the number of stages
  * Split into multiple chained methods

The overloads thing would essentially define fixed number of stages overloads like this:

```ts
async validateSequenceSuccess(docExists: boolean, stage1: TransactionStep, expected: object | undefined)
async validateSequenceSuccess(docExists: boolean, stage1: TransactionStep, stage2: TransactionStep, expected: object | undefined)
async validateSequenceSuccess(docExists: boolean, stage1: TransactionStep, stage2: TransactionStep, stage3: TransactionStep, expected: object | undefined)
```

This preserves this structure but changes the argument order. You'll need to unpack the arguments here in the usual way with typescript overloads. The approach works in the other languages, which all support proper overloads.

Splitting into chained methods will make the caller look like this:

```ts
tt.withDoc().run(getExisting, delete, delete).expectNoDoc();
tt.withDoc().run(getExisting, delete, update2).expectError('invalid-argument');
```

Note that `withDoc` and `run` don't actually do anything except assign into variables in the tester so they're not async and can return `this`. It's the expect methods that actually do everything.

I think the last form would read best but acknowledge that it's the most effort.